audiorecorder/src/test/java/org/odk/collect/audiorecorder/recording/MediaRecorderRecorderTest.kt

```diff
@@ -0,0 +1,224 @@
+
+import android.media.MediaRecorder
+import com.google.common.io.Files
+import org.hamcrest.MatcherAssert.assertThat
+import org.hamcrest.Matchers.equalTo
+import org.hamcrest.Matchers.not
+import org.junit.Test
+import org.odk.collect.audiorecorder.recorder.MediaRecorderRecorder
+import org.odk.collect.audiorecorder.recorder.MediaRecorderWrapper
+import java.io.File
+
+class MediaRecorderRecorderTest {
+
+    private val cacheDir = Files.createTempDir()
+    private val mediaRecorder = FakeMediaRecorderWrapper()
+    private val recorder = MediaRecorderRecorder(cacheDir) { mediaRecorder }
+
+    @Test
+    fun start_startsMediaRecorder() {
+        recorder.start()
+        assertThat(mediaRecorder.hasStarted(), equalTo(true))
+    }
+
+    @Test
+    fun start_setsUpAACRecordingFromMic() {
+        recorder.start()
+
+        assertThat(mediaRecorder.getAudioEncoder(), equalTo(MediaRecorder.AudioEncoder.AAC))
+        assertThat(mediaRecorder.getAudioEncodingSampleRate(), equalTo(32000))
+        assertThat(mediaRecorder.getAudioEncodingBitRate(), equalTo(128000))
+        assertThat(mediaRecorder.getOutputFormat(), equalTo(MediaRecorder.OutputFormat.MPEG_4))
+
+        assertThat(mediaRecorder.getAudioSource(), equalTo(MediaRecorder.AudioSource.MIC))
+    }
+
+    @Test
+    fun start_createsAndRecordsToFileInCacheDir() {
+        recorder.start()
+        assertThat(mediaRecorder.getOutputFile()!!.parent, equalTo(cacheDir.absolutePath))
+    }
+
+    @Test
+    fun recordingTwice_doesntUseSameOutputFile() {
+        var mediaRecorder = FakeMediaRecorderWrapper()
+        var recorder = MediaRecorderRecorder(cacheDir) { mediaRecorder }
+        recorder.start()
+        val outputFile1 = mediaRecorder.getOutputFile()
+
+        mediaRecorder = FakeMediaRecorderWrapper()
+        recorder = MediaRecorderRecorder(cacheDir) { mediaRecorder }
+        recorder.start()
+        val outputFile2 = mediaRecorder.getOutputFile()
+
+        assertThat(outputFile1!!.absolutePath, not(equalTo(outputFile2!!.absolutePath)))
+    }
+
+    @Test
+    fun stop_releasesMediaRecorder() {
+        recorder.start()
+        recorder.stop()
+        assertThat(mediaRecorder.isReleased(), equalTo(true))
+    }
+
+    @Test
+    fun stop_returnsOutputFile() {
+        recorder.start()
+        val file = recorder.stop()
+        assertThat(file.absolutePath, equalTo(mediaRecorder.getOutputFile()!!.absolutePath))
+    }
+
+    @Test
+    fun cancel_releasesMediaRecorder() {
+        recorder.start()
+        recorder.cancel()
+        assertThat(mediaRecorder.isReleased(), equalTo(true))
+    }
+
+    @Test
+    fun cancel_deletesOutputFile() {
+        recorder.start()
+        recorder.cancel()
+        assertThat(mediaRecorder.getOutputFile()!!.exists(), equalTo(false))
+    }
+
+    @Test
+    fun cancel_beforeStart_works() {
+        recorder.cancel()
+    }
+}
+
+private class FakeMediaRecorderWrapper : MediaRecorderWrapper {
```

Comment:
I went with a fake rather than a mock here as it lets me capture the importance of the method call order (`prepare` before `start` for example) without having to specify exactly what order things are called in.